src/Converter.cc

```diff
@@ -229,13 +240,270 @@ void Converter::ConvertImpl(tinyxml2::XMLElement *_elem,
     {
       Remove(_elem, childElem);
     }
+    else if (name == "unnest")
+    {
+      Unnest(_elem);
+      // TODO(jenn) delete debug statements
+      // std::cout << "\n\n_elem after unnest:\n"
+      //             << PrintElement(_elem) << std::endl;
+    }
     else if (name != "convert")
     {
       sdferr << "Unknown convert element[" << name << "]\n";
     }
   }
 }
 
+/////////////////////////////////////////////////
+void Converter::Unnest(tinyxml2::XMLElement *_elem)
+{
+  SDF_ASSERT(_elem != NULL, "SDF element is NULL");
+
+  tinyxml2::XMLDocument *doc = _elem->GetDocument();
+  tinyxml2::XMLNode *copy = DeepClone(doc, _elem);
+
+  // std::cout << "copy:\n" << PrintElement(copy->ToElement()) << std::endl;
+
+  tinyxml2::XMLElement *elem = _elem->FirstChildElement(), *nextElem = nullptr;
+  while (elem)
+  {
+    nextElem = elem->NextSiblingElement();
+
+    std::string elemName = elem->Name();
+
+    // skip element if not one of the following
+    if (elemName != "frame" && elemName != "joint"
+        && elemName != "link" && elemName != "model")
+    {
+      elem = nextElem;
+      continue;
+    }
+
+    std::string attrName = elem->Attribute("name");
+
+    size_t found = attrName.find("::");
+    if (found == std::string::npos)
+    {
+      // recursive unnest
+      if (elemName == "model")
+      {
+        // std::cout << "before recursive:\n"
+        //           << PrintElement(elem) << std::endl;
+        Unnest(elem);
+
+        // std::cout << "unnested model:\n" << PrintElement(elem) << std::endl;
+        _elem->InsertEndChild(elem);
+        break;
+      }
+
+      elem = nextElem;
+      continue;
+    }
+
+    std::string newModelName = attrName.substr(0, found);
+    tinyxml2::XMLElement *newModel = doc->NewElement("model");
+    newModel->SetAttribute("name", newModelName.c_str());
+
+    Converter::TupleVector deleteElems =
+        FindNewModelElements(doc,
+                            copy->ToElement(),
+                            newModel,
+                            found + 2);
+
+    // std::cout << "newModel:\n" << PrintElement(newModel) << std::endl;
+
+    if (!deleteElems.empty())
+    {
+      _elem->InsertEndChild(newModel);
+
+      // delete unnested elements
+      for (std::tuple<std::string, std::string> deleteElem : deleteElems)
+      {
+        tinyxml2::XMLElement *e =
+          _elem->FirstChildElement(std::get<0>(deleteElem).c_str());
+
+        std::string eName = std::get<1>(deleteElem);
+
+        while (eName != std::string(e->Attribute("name")))
+            e = e->NextSiblingElement(std::get<0>(deleteElem).c_str());
+
+        if (std::string(e->Attribute("name")) == eName)
+          _elem->DeleteChild(e);
+      }
+
+      nextElem = _elem->FirstChildElement();
```

Comment:
This seems like it will have to go through the whole model from the beginning every time a child model is unnested. Would it be possible to call `Unnest(newModel)` here and not have to start over?

Reply:
Yes, thanks for the tip a1006adfd84c3f908ed8f39630de3edc03098fb6

Reply:
I see that you've removed `DeepClone`. With that removed, and since the `InsertEndChild`s in `FindNewModelElements` move elements from their old location to their new location, I see the need for starting from the top. Performance wise, I think for most cases, this would be faster than doing `DeepClone`.